core/object/message_queue.cpp

```diff
@@ -35,94 +35,53 @@
-	Variant *v = memnew_placement(&buffer[buffer_end], Variant);
-	buffer_end += sizeof(Variant);
-	*v = p_value;
-
-	return OK;
+	page_messages[pages_used] = 0;
+	pages_used++;
+	page_offset = 0;
 }
 
-Error MessageQueue::push_notification(ObjectID p_id, int p_notification) {
-	_THREAD_SAFE_METHOD_
-
-	ERR_FAIL_COND_V(p_notification < 0, ERR_INVALID_PARAMETER);
-
-	uint8_t room_needed = sizeof(Message);
-
-	if ((buffer_end + room_needed) >= buffer_size) {
-		ERR_PRINT("Failed notification: " + itos(p_notification) + " target ID: " + itos(p_id) + ". Message queue out of memory. Try increasing \"memory/limits/message_queue/max_size_kb\" in project settings.");
-		statistics();
-		return ERR_OUT_OF_MEMORY;
-	}
-
-	Message *msg = memnew_placement(&buffer[buffer_end], Message);
-
-	msg->type = TYPE_NOTIFICATION;
-	msg->callable = Callable(p_id, CoreStringNames::get_singleton()->notification); //name is meaningless but callable needs it
-	//msg->target;
-	msg->notification = p_notification;
-
-	buffer_end += sizeof(Message);
-
-	return OK;
+Error CallQueue::push_callp(ObjectID p_id, const StringName &p_method, const Variant **p_args, int p_argcount, bool p_show_error) {
+	return push_callablep(Callable(p_id, p_method), p_args, p_argcount, p_show_error);
 }
 
-Error MessageQueue::push_callp(Object *p_object, const StringName &p_method, const Variant **p_args, int p_argcount, bool p_show_error) {
+Error CallQueue::push_callp(Object *p_object, const StringName &p_method, const Variant **p_args, int p_argcount, bool p_show_error) {
 	return push_callp(p_object->get_instance_id(), p_method, p_args, p_argcount, p_show_error);
 }
 
-Error MessageQueue::push_notification(Object *p_object, int p_notification) {
+Error CallQueue::push_notification(Object *p_object, int p_notification) {
 	return push_notification(p_object->get_instance_id(), p_notification);
 }
 
-Error MessageQueue::push_set(Object *p_object, const StringName &p_prop, const Variant &p_value) {
+Error CallQueue::push_set(Object *p_object, const StringName &p_prop, const Variant &p_value) {
 	return push_set(p_object->get_instance_id(), p_prop, p_value);
 }
 
-Error MessageQueue::push_callablep(const Callable &p_callable, const Variant **p_args, int p_argcount, bool p_show_error) {
-	_THREAD_SAFE_METHOD_
+Error CallQueue::push_callablep(const Callable &p_callable, const Variant **p_args, int p_argcount, bool p_show_error) {
+	mutex.lock();
+	uint32_t room_needed = sizeof(Message) + sizeof(Variant) * p_argcount;
 
-	int room_needed = sizeof(Message) + sizeof(Variant) * p_argcount;
+	_ensure_first_page();
 
-	if ((buffer_end + room_needed) >= buffer_size) {
-		ERR_PRINT("Failed method: " + p_callable + ". Message queue out of memory. Try increasing \"memory/limits/message_queue/max_size_kb\" in project settings.");
-		statistics();
-		return ERR_OUT_OF_MEMORY;
+	if ((page_offset + room_needed) > uint32_t(PAGE_SIZE_BYTES)) {
+		if (room_needed > uint32_t(PAGE_SIZE_BYTES) || pages_used == max_pages) {
```

Comment:
Am wondering, can any of these pushed messages require more than 4096 bytes (including parameters)?

Reply:
I think the 4096 bytes limit is sensible. Variants are normally 20-40 bytes depending on whether you compile floats to doubles, and messages as a base are about the same size. This means that the only way you  could exceed this is with a call with more than 100 arguments, which is probably unreasonable and unlikely.